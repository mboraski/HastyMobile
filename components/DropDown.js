// Third Party Imports
import React, { Component } from 'react';
import { 
    StyleSheet,
    View,
    TouchableOpacity,
    Animated
} from 'react-native';
import { connect } from 'react-redux';

// Relative Imports
import { emY } from '../utils/em';
import arrowIcon from '../assets/icons/disclosureIndicator.png';

class DropDown extends Component {
    state = {
        expanded: true,
        animation: new Animated.Value(),
        maxHeight: 0,
        minHeight: emY(3.3),
        spinValue: new Animated.Value(0),
    };

    componentDidMount() {
        const { minHeight } = this.state;
        this.state.animation.setValue(0);
        Animated.spring(
            this.state.animation,
            {
                toValue: minHeight
            }
        ).start();
        this.state.spinValue.setValue(1);
        Animated.timing(
            this.state.spinValue,
            {
                toValue: 1,
                duration: 400,
            }
        ).start();
        this.setState({ 
            expanded: false
        });
    }

    setMaxHeight = event => {
        this.setState({
            maxHeight: event.nativeEvent.layout.height
        });
    }
    
    setMinHeight = event => {
        this.setState({
            minHeight: event.nativeEvent.layout.height
        });
    }

    toggle = () => {
        const initialValue = (this.state.expanded) ? 
            this.state.maxHeight + this.state.minHeight : this.state.minHeight;
        const finalValue = (this.state.expanded) ? 
            this.state.minHeight : this.state.maxHeight + this.state.minHeight;
        const initialDegree = (this.state.expanded) ? 0 : 1;
        const finalDegree = (this.state.expanded) ? 1 : 0;
        this.setState({
            expanded: !this.state.expanded
        });
        this.state.animation.setValue(initialValue);
        Animated.spring(
            this.state.animation,
            {
                toValue: finalValue
            }
        ).start(); 
        this.state.spinValue.setValue(initialDegree);
        Animated.timing(
            this.state.spinValue,
            {
                toValue: finalDegree,
                duration: 400,
            }
        ).start();
    }

    render() {
        const { header } = this.props;
        const { animation } = this.state;
        const spin = this.state.spinValue.interpolate({
            inputRange: [0, 1],
            outputRange: ['0deg', '-90deg']
        });
        return (
            <View style={styles.container}>
                <Animated.View style={[styles.container, { height: animation }]}>
                    <View onLayout={this.setMinHeight}>
                        <TouchableOpacity onPress={this.toggle}>
                            {header}
                        </TouchableOpacity>
                    </View>
                    <View onLayout={this.setMaxHeight}>
                        {this.props.children}
                    </View>
                </Animated.View>
                <TouchableOpacity style={styles.arrowContainer} onPress={this.toggle}>
                    <Animated.Image 
                        source={arrowIcon}
                        style={[
                            arrowIcon,
                            {
                                transform: [{ rotate: spin }],
                                backgroundColor: 'rgba(0, 0, 0, 0)' 
                            }
                        ]}
                    />
                </TouchableOpacity>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        overflow: 'hidden'
    },
    arrowContainer: {
        position: 'absolute',
        right: 16,
        top: emY(1.5)
    },
    arrowIcon: {
        width: emY(0.9),
        height: emY(0.9)
    }
});

const mapDispatchToProps = function (dispatch) {
    return {};
};

export default connect(null, mapDispatchToProps)(DropDown);
