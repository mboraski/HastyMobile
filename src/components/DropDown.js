// Third Party Imports
import React, { Component } from 'react';
import { StyleSheet, View, TouchableOpacity, Animated } from 'react-native';

// Relative Imports
import { emY } from '../utils/em';
import arrowIcon from '../assets/icons/disclosureIndicator.png';

class DropDown extends Component {
    state = {
        expanded: true,
        animation: new Animated.Value(),
        maxHeight: 0,
        minHeight: emY(3.3),
        spinValue: new Animated.Value(0)
    };

    componentDidMount() {
        const { minHeight } = this.state;
        this.state.animation.setValue(0);
        Animated.spring(this.state.animation, {
            toValue: minHeight
        }).start();
        this.state.spinValue.setValue(1);
        Animated.timing(this.state.spinValue, {
            toValue: 1,
            duration: 400
        }).start();
        this.setState({
            expanded: false
        });
    }

    setMaxHeight = event => {
        this.setState({
            maxHeight: event.nativeEvent.layout.height
        });
    };

    setMinHeight = event => {
        this.setState({
            minHeight: event.nativeEvent.layout.height
        });
    };

    toggle = () => {
        const initialValue = this.state.expanded
            ? this.state.maxHeight + this.state.minHeight
            : this.state.minHeight;
        const finalValue = this.state.expanded
            ? this.state.minHeight
            : this.state.maxHeight + this.state.minHeight;
        const initialDegree = this.state.expanded ? 0 : 1;
        const finalDegree = this.state.expanded ? 1 : 0;
        this.setState({
            expanded: !this.state.expanded
        });
        this.state.animation.setValue(initialValue);
        Animated.spring(this.state.animation, {
            toValue: finalValue
        }).start();
        this.state.spinValue.setValue(initialDegree);
        Animated.timing(this.state.spinValue, {
            toValue: finalDegree,
            duration: 400
        }).start();
    };

    render() {
        const { header } = this.props;
        const { animation } = this.state;
        const spin = this.state.spinValue.interpolate({
            inputRange: [0, 1],
            outputRange: ['0deg', '-90deg']
        });
        return (
            <View style={styles.container}>
                <Animated.View
                    style={[styles.container, { height: animation }]}
                >
                    <View onLayout={this.setMinHeight}>
                        <TouchableOpacity onPress={this.toggle}>
                            {header}
                            <View style={styles.arrowContainer}>
                                <Animated.Image
                                    source={arrowIcon}
                                    style={[
                                        styles.arrowIcon,
                                        {
                                            transform: [{ rotate: spin }]
                                        }
                                    ]}
                                />
                            </View>
                        </TouchableOpacity>
                    </View>
                    <View onLayout={this.setMaxHeight}>
                        {this.props.children}
                    </View>
                </Animated.View>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        overflow: 'hidden',
        backgroundColor: 'white'
    },
    arrowContainer: {
        position: 'absolute',
        right: 16,
        top: 0,
        bottom: 0,
        alignItems: 'center',
        justifyContent: 'center'
    },
    arrowIcon: {
        width: emY(0.9),
        height: emY(0.9)
    }
});

export default DropDown;
