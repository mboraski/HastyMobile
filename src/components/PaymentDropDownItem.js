// Third Party Imports
import React, { Component } from 'react';
import {
    StyleSheet,
    Image,
    TouchableOpacity,
} from 'react-native';
import { connect } from 'react-redux';

// Relative Imports
import Text from './Text';
import Color from '../constants/Color';
import { emY } from '../utils/em';
import mastercard from '../assets/icons/master-card.png';
import visa from '../assets/icons/visa.png';
// import arrowIcon from '../assets/icons/disclosureIndicator.png';
import { selectCard as onPress } from '../actions/paymentActions';

const SIZE = emY(2);

type Props = {
    isHeaderItem: boolean
};

class PaymentDropDownItem extends Component {
    props: Props;

    onSelectCard = () => {
        onPress(this.props.card);
    }

    render() {
        const { isHeaderItem, card } = this.props;
        const { last4, brand } = card;
        // const arrowIconMark = isHeaderItem ?
        //     (<Image source={arrowIcon} style={styles.arrowIcon} />) :
        //     null;
        let paymentIcon = null;
        if (brand === 'Visa') {
            paymentIcon = visa;
        } else if (brand === 'MasterCard') {
            paymentIcon = mastercard;
        }
        const color = isHeaderItem ?
            Color.YELLOW_600 : 'transparent';
        return (
            <TouchableOpacity
                onPress={this.onSelectCard}
                style={[styles.container, { borderColor: color }]}
                disabled={isHeaderItem}
            >
                <Image
                    source={paymentIcon}
                    style={styles.paymentImage}
                    resizeMode="contain"
                />
                <Text style={styles.paymentNumber}>{last4}</Text>
                {/* {arrowIconMark} */}
            </TouchableOpacity>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        paddingHorizontal: 16,
        paddingVertical: emY(0.89),
        backgroundColor: Color.GREY_100,
        borderColor: Color.YELLOW_600,
        borderLeftWidth: 0,
        borderRightWidth: 0,
        borderTopWidth: 1,
        borderBottomWidth: 1
    },
    paymentImage: {
        width: emY(2.88),
        height: emY(1.38),
    },
    paymentNumber: {
        fontSize: emY(1.08),
        marginLeft: 10
    },

});

const mapDispatchToProps = {
    onPress
};

export default connect(null, mapDispatchToProps)(PaymentDropDownItem);
