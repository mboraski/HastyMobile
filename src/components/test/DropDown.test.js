import React from 'react';
import 'react-native';
import { shallow } from 'enzyme';
import DropDown from '../DropDown';
import PaymentDropDownItem from '../PaymentDropDownItem';

describe('DropDown', () => {
    it('renders correctly', () => {
        const render = shallow(
            <DropDown header={<PaymentDropDownItem isHeaderItem />}>
                <PaymentDropDownItem isHeaderItem={false} />
                <PaymentDropDownItem isHeaderItem={false} />
            </DropDown>
        );
        expect(render).toMatchSnapshot();
    });
});
