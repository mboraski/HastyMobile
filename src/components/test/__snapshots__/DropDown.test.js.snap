// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown renders correctly 1`] = `
<View
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <AnimatedComponent
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "height": 0,
        },
      ]
    }
  >
    <View
      onLayout={[Function]}
    >
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
      >
        <Connect(PaymentDropDownItem)
          isHeaderItem={true}
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "bottom": 0,
              "justifyContent": "center",
              "position": "absolute",
              "right": 16,
              "top": 0,
            }
          }
        >
          <AnimatedComponent
            source={1}
            style={
              Array [
                Object {
                  "height": 14.4,
                  "width": 14.4,
                },
                Object {
                  "transform": Array [
                    Object {
                      "rotate": "-90deg",
                    },
                  ],
                },
              ]
            }
          />
        </View>
      </TouchableOpacity>
    </View>
    <View
      onLayout={[Function]}
    >
      <Connect(PaymentDropDownItem)
        isHeaderItem={false}
      />
      <Connect(PaymentDropDownItem)
        isHeaderItem={false}
      />
    </View>
  </AnimatedComponent>
</View>
`;
